extensions/pkg/predicate/preconditions.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2022 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package predicate
+
+import (
+	"context"
+
+	extensionscontroller "github.com/gardener/gardener/extensions/pkg/controller"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	contextutil "github.com/gardener/gardener/pkg/utils/context"
+
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+)
+
+// IsInGardenNamespacePredicate is a predicate which returns true when the provided object is in the 'garden' namespace.
+var IsInGardenNamespacePredicate = predicate.NewPredicateFuncs(func(obj client.Object) bool {
+	return obj != nil && obj.GetNamespace() == v1beta1constants.GardenNamespace
+})
+
+// ShootNotFailedPredicate returns a predicate which returns true when the Shoot's `.status.lastOperation.state` is not
+// equals 'Failed'.
+func ShootNotFailedPredicate() predicate.Predicate {
+	return &shootNotFailedPredicate{}
+}
+
+type shootNotFailedPredicate struct {
+	ctx   context.Context
+	cache cache.Cache
+}
+
+func (p *shootNotFailedPredicate) InjectFunc(f inject.Func) error {
+	return f(p)
+}
```

Comment:
I guess, this will cause an endless recursion. `InjectFunc` is for relaying the injection logic to some member, e.g. when nesting reconcilers or predicates. We don't need it here, so it should be removed.